### PR TITLE
Doc sphinx

### PR DESCRIPTION
**Pull request by pthom**
Hello,

This PR adds several things:

**The first commit concentrates on the code layout**

* It provides two libraries: one with a merged headers layout convention, and one that uses the top include folder. In the merged header convention, a private/ subfolder is demonstrated.
* It demonstrates a layout with two levels of subfolders, which is common
````bash
include
└── Mamals/
    └── Felinae/
        └── cat.hpp

src
├── Mamals/
│   ├── Canidae/                   # a library that uses merged header placement
│   │   ├── dog.cpp
│   │   ├── dog.h
│   │   ├── dog.test.cpp
│   │   └── private/              # with private headers
│   │       └── dog_internal.h
│   ├── Felinae/                    # a library that uses the top include/ folder
│   │   ├── cat.cpp
│   │   └── cat.test.cpp
│   └── Mamals.h
└── main.test.cpp
```` 
* It demonstrates that even in the case of merged header placement, includes can use the two parent folders
* It demonstates that a wise choice is to say that any folder should introduce a namespace with same name (with no variation). This is the sane strategy used by python where a namespace is added automatically whenever a folder is encountered. 
For example, with `dog.cpp`:
````cpp
#include "Mamals/Canidae/dog.h"

namespace Mamals {
namespace Canidae {
std::string Dog::bark() { return "Wof"; };
}
}
````

**The second commit adds sophisticated doc generator tooling, and the third adds the related github actions**

This tooling is based on Doxygen, sphinx, breathe, and readthedocs. This is the tooling used by  [fmtlib](https://fmt.dev/latest/index.html). It provides quite nice docs. See the [output](https://pthom-cpp-project-template.readthedocs.io/en/latest/). . 

The plus side is that this kind of doc is extremely versatile (and can be extended to add diagrams via graphviz and plantuml, math equations, etc.)
The downside is that the doc should written using rst files, and that writing it requires more learning. 

May be this is too much for a starter template. Anyhow, I worked on it because I am interested in having such a doc in one of my projects.

As a matter of fact, these two commit are quite independent. They could be decoupled if you want.  They are also quite opinionated, so feel free to accept them or not :-)

### CODE REVIEW

**Comment by pthom** [comment]
Wait a bit, I need to correct the CI (and add a CI step for the doc)

**Comment by pthom** [comment]
Ok, the CI now passes and builds the doc: see the [CI logs](https://github.com/KaiSzuttor/cpp_project_template/pull/2/checks?check_run_id=682585145)

**Comment by KaiSzuttor** [comment]
Thanks a lot for your effort... it will take me some time to go through your proposed changes :)

**Comment by KaiSzuttor** [comment]
> It provides two libraries: one with a merged headers layout convention, and one that uses the top include folder. In the merged header convention, a private/ subfolder is demonstrated.

I think in the pitchfork layout convention there is no distinguishing between public and private headers if a merged header convention is used, at least I could not find any hint on this. I think the most clean way is to either use separate header placement or not care about private headers at all, what do you think?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -28,9 +28,18 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@master
+    - name: Install Doxygen
```
Comment (KaiSzuttor) [inline .github/workflows/build.yml:31]: I wanted to look at using a pre-build docker image within github actions... If there are more and more dependencies you don't want to reinstall all those for each pull request. I have to look into that.

**Comment by KaiSzuttor** [comment]
> This tooling is based on Doxygen, sphinx, breathe, and readthedocs

Is the amount of complication paying off here? A plain doxygen documentation can already be searched through, mathematical notation and stuff like that is also possible... Could you pinpoint me to the advantages of this combination?

**Comment by KaiSzuttor** [comment]
> It demonstates that a wise choice is to say that any folder should introduce a namespace with same name (with no variation). This is the sane strategy used by python where a namespace is added automatically whenever a folder is encountered.

Isn't this already in the repository for ```cat/sounds``` -> ```Cat::Sounds```?

**Comment by KaiSzuttor** [comment]
> As a matter of fact, these two commit are quite independent. They could be decoupled if you want. They are also quite opinionated, so feel free to accept them or not :-)

I'm open for discussion and I hope you don't mind if I may drop some of your suggestions 😄

File: .github/workflows/build.yml (hunk 1 above)
Comment (pthom) [reply .github/workflows/build.yml:31]: ok

**Comment by pthom** [comment]
> I think in the pitchfork layout convention there is no distinguishing between public and private headers if a merged header convention is used, at least I could not find any hint on this. I think the most clean way is to either use separate header placement or not care about private headers at all, what do you think?

I think that with the [upcoming modules in C++20](https://en.cppreference.com/w/cpp/language/modules), the separate header layout convention will become obsolete for projects that use modules. 
However, it is still important to distinguish between the "published" API of a library, i.e the API the other projects may use, and its "internal" api. This is the reasoning behind the private folder; and this is why I think that "internal" might be a better name.

> Is the amount of complication paying off here? A plain doxygen documentation can already be searched through, mathematical notation and stuff like that is also possible... Could you pinpoint me to the advantages of this combination?

The advantage is a [more beautiful doc](https://pthom-cpp-documentation-example.readthedocs.io/en/latest/), where you can mix general hand-written documentation (see [index.rst](https://github.com/pthom/cpp_project_template/blob/doc_puml/docs/index.rst) ) with automated documentation by doxygen.

> Isn't this already in the repository for cat/sounds -> Cat::Sounds?

You are right. However, a small typo: cat.cpp does not include cat.hpp 

> I'm open for discussion and I hope you don't mind if I may drop some of your suggestions 😄

Do as you wish, I totally understand that this kind of doc is more powerful, but that there is learning curve at the beginning that may not be worth it. 
A solution might be to have a separate branch for this. I will keep the branch in my fork anyhow ;-)

File: .github/workflows/build.yml (hunk 1 above)
Comment (KaiSzuttor) [reply .github/workflows/build.yml:31]: update on that: https://github.community/t/github-actions-runs-on-custom-docker-image/116846/2

so it seems like this is not yet possible (at least we cannot use images from the github package registry (yet))

**Comment by KaiSzuttor** [comment]
sorry, for no action for long time. I really appreciate your contribution but want to keep this template as basic as possible. So I think your changes are more like an extension to the basic template and not as generic as I want the template to be.

**Pull request closed without merging**